File: Assets/Scripts/UI/AccessibilityBridge.cs
using System;
using UnityEngine;

namespace UnityMobileQA.UI
{
    /// <summary>
    /// Bridges Unity AccessibilityLabel to native contentDescription (Android) / accessibilityIdentifier (iOS).
    /// On Android: Sets root view contentDescription via native plugin. Per-element requires Unity 6 Accessibility module.
    /// On iOS: accessibilityIdentifier requires Unity 6 or native plugin; gameObject.name used as fallback.
    /// </summary>
    public static class AccessibilityBridge
    {
        private const string AndroidBridgeClass = "com.unitymobileqa.accessibility.AccessibilityBridge";

        /// <summary>Call at app start to set root contentDescription (Android). Improves Appium discovery.</summary>
        public static void SetRootContentDescription(string description)
        {
#if UNITY_ANDROID && !UNITY_EDITOR
            try
            {
                using (var bridge = new AndroidJavaClass(AndroidBridgeClass))
                {
                    bridge.CallStatic("setRootContentDescription", description ?? "UnityMobileQA");
                }
            }
            catch (Exception e)
            {
                Debug.LogWarning($"[AccessibilityBridge] setRootContentDescription failed: {e.Message}");
            }
#endif
        }
    }
}

File: Assets/Plugins/Android/AccessibilityBridge.java
package com.unitymobileqa.accessibility;

import android.app.Activity;
import android.view.View;
import com.unity3d.player.UnityPlayer;

/**
 * Native bridge for accessibility labels. Sets contentDescription on the root view
 * so Appium/UiAutomator can identify the app. For per-element contentDescription,
 * Unity 6 Accessibility module or UAP plugin is required (Unity 2022 renders to single view).
 */
public class AccessibilityBridge {
    /**
     * Set contentDescription on the root view. Call from Unity via AndroidJavaObject.
     * Enables Appium to find the app; per-element IDs require Unity 6.
     */
    public static void setRootContentDescription(String description) {
        try {
            Activity activity = UnityPlayer.currentActivity;
            if (activity != null) {
                View root = activity.getWindow().getDecorView().getRootView();
                if (root != null) {
                    root.setContentDescription(description);
                }
            }
        } catch (Exception ignored) {
            // Silently fail if not on Android or view not ready
        }
    }
}
